src/routes/github/github-oauth-router.ts

```diff
@@ -205,8 +206,45 @@ const getCloudOrGHESAppClientSecret = async (gitHubAppConfig, jiraHost: string)
 	return ghesApp.getDecryptedGitHubClientSecret(jiraHost);
 };
 
+const TempGetJiraHostFromStateMiddleware  = async (req: Request, res: Response, next: NextFunction) => {
+
+	const stateKey = req.query.state as string;
+	if (!stateKey) {
+		req.log.warn("State key is empty");
+		res.status(400).send(Errors.MISSING_JIRA_HOST);
+		return;
+	}
+	if (!req.session[stateKey]) {
+		req.log.warn("State is empty in req.session for callback redirect");
+		res.status(400).send(Errors.MISSING_JIRA_HOST);
+		return;
+	}
+
+	let url;
+	try {
+		let redirectUrl = req.session[stateKey];
+		if (!redirectUrl.startsWith("https")) {
+			redirectUrl = "https://dummy.domain" + redirectUrl;
+		}
+		url = new URL(redirectUrl);
+	} catch (e) {
+		req.log.warn("Error passing redirect url in callback", e);
+		res.status(400).send(Errors.MISSING_JIRA_HOST);
+		return;
+	}
+	const jiraHost = url.searchParams.get("jiraHost");
+	if (!jiraHost) {
+		req.log.warn("jiraHost is missing in state redirect uri");
+		res.status(400).send(Errors.MISSING_JIRA_HOST);
+		return;
+	}
+
+	res.locals.jiraHost = jiraHost;
+	next();
+};
+
 // IMPORTANT: We need to keep the login/callback/middleware functions
 // in the same file as they reference each other
-export const GithubOAuthRouter = Router();
-GithubOAuthRouter.get("/login", GithubOAuthLoginGet);
-GithubOAuthRouter.get(callbackSubPath, GithubOAuthCallbackGet);
+export const GithubOAuthRouter = Router({ mergeParams: true });
+GithubOAuthRouter.get("/login", GithubServerAppMiddleware, GithubOAuthLoginGet);
+GithubOAuthRouter.get(callbackSubPath, TempGetJiraHostFromStateMiddleware, GithubServerAppMiddleware, GithubOAuthCallbackGet);
```

Comment:
Only `create-branch` flow contains `jirahost` as a query parameter. Other flows which don't have `jiraHost` query parameter will fail bcz of `TempGetJiraHostFromStateMiddleware`

Reply:
This callback originated from the login get, which supposed to have a jiraHost in the session[stateKey]. So I think it is always there?

Reply:
You are right, let me pick the jiraHost from session if available

Reply:
I doubt for GH cloud or enterpise flows. Have you tested it ?

Reply:
Okay I fixed the bug.